File: sandbox-runner/patterns/cancellation-propagation/typescript/starter.ts
import { Client, Connection } from "@temporalio/client";

import { TASK_QUEUE, WORKFLOW_ID_PREFIX } from "./shared";
import { fulfillOrderWorkflow, stopSignal } from "./workflows";

async function main(): Promise<void> {
  const connection = await Connection.connect();
  try {
    const client = new Client({ connection });
    const workflowId = `${WORKFLOW_ID_PREFIX}-${Date.now()}`;
    const orderId = "order-42";

    const handle = await client.workflow.start(fulfillOrderWorkflow, {
      args: [orderId],
      taskQueue: TASK_QUEUE,
      workflowId,
    });
    console.log(`Started workflow: ${workflowId}`);
    console.log(`Fulfilling ${orderId}; children are reserving resources concurrently…`);

    // Let the children apply their steps, then request a stop.
    await new Promise((r) => setTimeout(r, 2000));
    console.log("Requesting stop; cancellation will propagate to every child…");
    await handle.signal(stopSignal);

    const result = await handle.result();
    console.log(result);
    console.log(
      `Open the Temporal UI and search for '${workflowId}' to see each child transition to Canceled after compensating.`
    );
  } finally {
    await connection.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: sandbox-runner/patterns/cancellation-propagation/python/starter.py
import asyncio
import time

from temporalio.client import Client

from shared import TASK_QUEUE, WORKFLOW_ID_PREFIX
from workflows import FulfillOrderWorkflow


async def main() -> None:
    client = await Client.connect("localhost:7233")
    workflow_id = f"{WORKFLOW_ID_PREFIX}-{int(time.time() * 1000)}"
    order_id = "order-42"

    handle = await client.start_workflow(
        FulfillOrderWorkflow.run,
        order_id,
        id=workflow_id,
        task_queue=TASK_QUEUE,
    )
    print(f"Started workflow: {workflow_id}")
    print(f"Fulfilling {order_id}; children are reserving resources concurrently…")

    # Let the children apply their steps, then request a stop.
    await asyncio.sleep(2)
    print("Requesting stop; cancellation will propagate to every child…")
    await handle.signal(FulfillOrderWorkflow.stop)

    result = await handle.result()
    print(result)
    print(
        f"Open the Temporal UI and search for '{workflow_id}' to see each child "
        f"transition to Canceled after compensating."
    )


if __name__ == "__main__":
    asyncio.run(main())
